File: openinstall-ios/DeviceFingerprint.swift
//
//  DeviceFingerprint.swift
//  OpenInstall iOS SDK
//
//  设备指纹收集
//

import UIKit
import Foundation

class DeviceFingerprint {
    
    /// 收集设备指纹
    static func collect() -> [String: Any] {
        let device = UIDevice.current
        let screen = UIScreen.main
        
        var fingerprint: [String: Any] = [
            "userAgent": getUserAgent(),
            "platform": "iOS",
            "osVersion": device.systemVersion,
            "deviceModel": device.model,
            "deviceName": device.name,
            "screenWidth": Int(screen.bounds.width * screen.scale),
            "screenHeight": Int(screen.bounds.height * screen.scale),
            "screenScale": screen.scale,
            "timezone": TimeZone.current.identifier,
            "timezoneOffset": TimeZone.current.secondsFromGMT() / 60, // 转换为分钟
            "language": Locale.current.languageCode ?? ""
        ]
        
        return fingerprint
    }
    
    /// 获取 User Agent
    static func getUserAgent() -> String {
        let device = UIDevice.current
        return "iOS/\(device.systemVersion) \(device.model)"
    }
    
    /// 生成设备指纹ID
    static func generateId(from fingerprint: [String: Any]) -> String {
        guard let jsonData = try? JSONSerialization.data(withJSONObject: fingerprint, options: []),
              let jsonString = String(data: jsonData, encoding: .utf8) else {
            return ""
        }
        
        // 使用简单的哈希算法
        var hash = 0
        for char in jsonString.utf8 {
            hash = ((hash << 5) - hash) + Int(char)
            hash = hash & hash // Convert to 32bit integer
        }
        
        return String(abs(hash), radix: 36)
    }
}

File: openinstall-backend/src/main/java/com/openinstall/model/DeviceFingerprint.java
package com.openinstall.model;

import com.fasterxml.jackson.annotation.JsonInclude;
import lombok.AllArgsConstructor;
import lombok.Data;
import lombok.NoArgsConstructor;

import java.util.Map;

/**
 * 设备指纹信息
 */
@Data
@NoArgsConstructor
@AllArgsConstructor
@JsonInclude(JsonInclude.Include.NON_NULL)
public class DeviceFingerprint {
    
    private String userAgent;
    private String language;
    private String platform;
    private Integer screenWidth;
    private Integer screenHeight;
    private Integer screenColorDepth;
    private Double pixelRatio;
    private String timezone;
    private Integer timezoneOffset;
    private String canvasFingerprint;
    private Map<String, Object> webglFingerprint;
    private Boolean cookieEnabled;
    private String doNotTrack;
    
    // iOS/Android 特有字段
    private String osVersion;
    private String deviceModel;
    private String deviceBrand;
    private String deviceName;
    private Double screenScale;
    private Double screenDensity;
}

